### ISSUE TITLE
Another question for help

### ISSUE THREAD

**Issue by willinglion** [zcakhaa/DeepLOB-Deep-Convolutional-Neural-Networks-for-Limit-Order-Books#5]
'The labels are then decided based on a threshold (a) for the percentage change (lt).' What is the method which can be used to set the value of the threshold?

**Comment by zcakhaa** [comment]
l_t is defined in Equation 4 in the paper. You can set a \alpha and then use l_t to decide labels.
[cid:202bbf1f-7993-4b2a-82fb-267ee335cd64]

________________________________

**Comment by willinglion** [comment]
l_t has its method in the paper. but a \alpha I do not know how to set it. I set several a \alpha by man-made.  the results of accuracy is far away from them in your paper.  How can I set  a \alpha? Must I loop through all possible values of a \alpha to see which one has the best results?

**Comment by zcakhaa** [comment]
The choice of \alpha depends on your application. For academic purposes, we set \alpha for each stock to have a balanced training set. In practice, you can set it to a higher threshold to trade only if the movement is great. I don't think you should loop through all \alpha to just get good results.

________________________________

**Comment by willinglion** [comment]
what you mean 'have a balanced training set' is let the proportion(1,0,-1) to be 1:1:1?

**Comment by zcakhaa** [comment]
Yes, each class accounts for about 33% in the training data

________________________________

**Comment by willinglion** [comment]
Great, thank you. higher thresholds can help Trading Strategy get more profit. I have try  some higher thresholds, and it will make the class aacount(0) higher, such as 20%:60%:20%. The model can learn the accuracy more than 60%. but the accuracy is mainly for the class aacount(0). it is very low for the class aacounts(-1,1). it is get a bad profit for Trading Strategy, only class aacounts(-1,1) making trade. So i think the imbalance of class aacounts will let the model always select 0 for classification and get worse result comparing with mean proportion 33%.  it is confuse me how to make the model get a better result by higher thresholds. Can you give me some suggestion?

**Comment by zcakhaa** [comment]
You can check our paper on discussing trading rules: https://arxiv.org/abs/1811.10041. Not necessarily related to label parameter \alpha, but it would be helpful. A higher threshold is needed when you take transaction costs into account and you can set \alpha to your cost, so you know at least the movement is greater than your cost.


________________________________

**Comment by willinglion** [comment]
Thanks for you help. I really need a higher threshold, as the transaction cost is not free.

**Comment by willinglion** [comment]
I have read your paper 'BDLOB: Bayesian Deep Convolutional Neural Networks for Limit Order Books'. The solution with Bayesian networks is so wonderful. I do not find code for Bayesian network in your project. 
'we add a we apply a dropout layer with rate 0.2 after the Inception Module, decided using grid-search methods'. Is it the code as following?
# use the MC dropout here
    conv_reshape = Reshape((int(convsecond_output.shape[1]), int(convsecond_output.shape[3])))(convsecond_output)

**Comment by zcakhaa** [comment]
Code is not published, but it is very simple to implement. You just need to use dropout and enable it during the testing period.


________________________________

**Comment by willinglion** [comment]
I add a dropout layer with rate 0.2 after the Inception Module and applied Concrete Dropout to
convolutional layers in the Inception Module, by import the class ConcreteDropout(Wrapper) in the paper Concrete Dropout, as the paper introduce. The code is as following, but it ovefit after several epochs. Can you give me some suggestion about my implement code of your paper?
# build the inception module
    convsecond_1 = ConcreteDropout(Conv2D(64, (1, 1), padding='same'))(conv_first1)
    convsecond_1 = keras.layers.LeakyReLU(alpha=0.01)(convsecond_1)
    convsecond_1 = ConcreteDropout(Conv2D(64, (3, 1), padding='same'))(convsecond_1)
    convsecond_1 = keras.layers.LeakyReLU(alpha=0.01)(convsecond_1)

    convsecond_2 = ConcreteDropout(Conv2D(64, (1, 1), padding='same'))(conv_first1)
    convsecond_2 = keras.layers.LeakyReLU(alpha=0.01)(convsecond_2)
    convsecond_2 = ConcreteDropout(Conv2D(64, (5, 1), padding='same'))(convsecond_2)
    convsecond_2 = keras.layers.LeakyReLU(alpha=0.01)(convsecond_2)

    convsecond_3 = MaxPooling2D((3, 1), strides=(1, 1), padding='same')(conv_first1)
    convsecond_3 = ConcreteDropout(Conv2D(64, (1, 1), padding='same'))(convsecond_3)
    convsecond_3 = keras.layers.LeakyReLU(alpha=0.01)(convsecond_3)
    
    convsecond_output = keras.layers.concatenate([convsecond_1, convsecond_2, convsecond_3], axis=3)

    # use the MC dropout here
    convsecond_output_drop = Dropout(0.2)(convsecond_output)
    conv_reshape = Reshape((int(convsecond_output.shape[1]), int(convsecond_output.shape[3])))(convsecond_output_drop)

    # build the last LSTM layer
    conv_lstm = CuDNNLSTM(number_of_lstm)(conv_reshape)

**Comment by zcakhaa** [comment]
You need to enable the dropout during testing, otherwise the dropout is switched off. Also, you need to use mask to drop the entire sequence of a feature map from the Inception module.


________________________________

**Comment by HaishuoFang** [comment]
> Great, thank you. higher thresholds can help Trading Strategy get more profit. I have try some higher thresholds, and it will make the class aacount(0) higher, such as 20%:60%:20%. The model can learn the accuracy more than 60%. but the accuracy is mainly for the class aacount(0). it is very low for the class aacounts(-1,1). it is get a bad profit for Trading Strategy, only class aacounts(-1,1) making trade. So i think the imbalance of class aacounts will let the model always select 0 for classification and get worse result comparing with mean proportion 33%. it is confuse me how to make the model get a better result by higher thresholds. Can you give me some suggestion?

I have the same problem, do you try to downsample class 0 to make balanced data?

**Comment by zcakhaa** [comment]
Hi, in general, I don't downsample the data to make it balance, but I guess you can always try and compare the results. Otherwise, you can use some weighted loss function to correct this.

________________________________

**Comment by YoussefElFarahaty21** [comment]
can you upload the model weights?
